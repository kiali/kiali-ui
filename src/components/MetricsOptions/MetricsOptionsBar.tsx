import * as React from 'react';
import { Toolbar, ToolbarRightContent, DropdownButton, MenuItem, Icon } from 'patternfly-react';

import ValueSelectHelper from './ValueSelectHelper';
import MetricsOptions from '../../types/MetricsOptions';
import RateIntervals from '../../types/RateIntervals';

interface Props {
  onOptionsChanged: (opts: MetricsOptions) => void;
  onPollIntervalChanged: (interval: number) => void;
  loading?: boolean;
}

interface MetricsOptionsState {
  rateInterval: string;
  pollInterval: number;
  duration: number;
  ticks: number;
  groupByLabels: string[];
}

interface GroupByLabel {
  labelIn: string;
  labelOut: string;
}

export class MetricsOptionsBar extends React.Component<Props, MetricsOptionsState> {
  static DefaultRateInterval = RateIntervals[0][0];

  static PollIntervals = [
    [0, 'Pause'],
    [5000, '5 seconds'],
    [10000, '10 seconds'],
    [30000, '30 seconds'],
    [60000, '1 minute'],
    [300000, '5 minutes']
  ];
  static DefaultPollInterval = 5000;

  static Ticks = [10, 20, 30, 50, 100, 200];
  static DefaultTicks = MetricsOptionsBar.Ticks[2];

  static Durations: [number, string][] = [
    [300, 'Last 5 minutes'],
    [600, 'Last 10 minutes'],
    [1800, 'Last 30 minutes'],
    [3600, 'Last hour'],
    [10800, 'Last 3 hours'],
    [21600, 'Last 6 hours'],
    [43200, 'Last 12 hours'],
    [86400, 'Last day']
  ];
  static DefaultDuration = MetricsOptionsBar.Durations[1][0];

  static GroupByLabelOptions: { [key: string]: GroupByLabel } = {
    'local version': {
      labelIn: 'destination_version',
      labelOut: 'source_version'
    },
    'remote service': {
      labelIn: 'source_service',
      labelOut: 'destination_service'
    },
    'remote version': {
      labelIn: 'source_version',
      labelOut: 'destination_version'
    },
    'response code': {
      labelIn: 'response_code',
      labelOut: 'response_code'
    }
  };

  groupByLabelsHelper: ValueSelectHelper;

  constructor(props: Props) {
    super(props);

    this.onRateIntervalChanged = this.onRateIntervalChanged.bind(this);
    this.onPollIntervalChanged = this.onPollIntervalChanged.bind(this);
    this.onDurationChanged = this.onDurationChanged.bind(this);
    this.onTicksChanged = this.onTicksChanged.bind(this);
    this.changedGroupByLabel = this.changedGroupByLabel.bind(this);

    this.groupByLabelsHelper = new ValueSelectHelper({
      items: Object.keys(MetricsOptionsBar.GroupByLabelOptions),
      onChange: this.changedGroupByLabel,
      dropdownTitle: 'Group by',
      resultsTitle: 'Grouping by:'
    });

    this.state = {
      rateInterval: MetricsOptionsBar.DefaultRateInterval,
      pollInterval: MetricsOptionsBar.DefaultPollInterval,
      duration: MetricsOptionsBar.DefaultDuration,
      ticks: MetricsOptionsBar.DefaultTicks,
      groupByLabels: []
    };
  }

  componentDidMount() {
    // Init state upstream
    this.reportOptions();
    this.props.onPollIntervalChanged(this.state.pollInterval);
  }

  onRateIntervalChanged(key: string) {
    this.setState({ rateInterval: key }, () => {
      this.reportOptions();
    });
  }

  onPollIntervalChanged(key: number) {
    // We use a specific handler so that changing poll interval doesn't trigger a metrics refresh in parent
    // Especially useful when pausing
    this.props.onPollIntervalChanged(key);
    this.setState({ pollInterval: key });
  }

  onDurationChanged(key: number) {
    this.setState({ duration: key }, () => {
      this.reportOptions();
    });
  }

  onTicksChanged(key: number) {
    this.setState({ ticks: key }, () => {
      this.reportOptions();
    });
  }

  reportOptions() {
    // State-to-options converter (removes unnecessary properties)
    const labelsIn = this.state.groupByLabels.map(lbl => MetricsOptionsBar.GroupByLabelOptions[lbl].labelIn);
    const labelsOut = this.state.groupByLabels.map(lbl => MetricsOptionsBar.GroupByLabelOptions[lbl].labelOut);
    this.props.onOptionsChanged({
      rateInterval: this.state.rateInterval,
      rateFunc: 'irate',
      duration: this.state.duration,
      step: this.state.duration / this.state.ticks,
      byLabelsIn: labelsIn,
      byLabelsOut: labelsOut
    });
  }

  changedGroupByLabel(labels: string[]) {
    this.setState({ groupByLabels: labels }, () => {
      this.reportOptions();
    });
  }

  render() {
    return (
      <Toolbar>
        {this.groupByLabelsHelper.renderDropdown()}
        <div className="form-group">
          <DropdownButton id="duration" title="Duration" onSelect={this.onDurationChanged}>
            {MetricsOptionsBar.Durations.map(r => (
              <MenuItem key={r[0]} active={r[0] === this.state.duration} eventKey={r[0]}>
                {r[1]}
              </MenuItem>
            ))}
          </DropdownButton>
          <DropdownButton id="ticks" title="Ticks" onSelect={this.onTicksChanged}>
            {MetricsOptionsBar.Ticks.map(r => (
              <MenuItem key={r} active={r === this.state.ticks} eventKey={r}>
                {r}
              </MenuItem>
            ))}
          </DropdownButton>
          <DropdownButton id="rateInterval" title="Rate interval" onSelect={this.onRateIntervalChanged}>
            {RateIntervals.map(r => (
              <MenuItem key={r[0]} active={r[0] === this.state.rateInterval} eventKey={r[0]}>
                {r[1]}
              </MenuItem>
            ))}
          </DropdownButton>
          <DropdownButton id="pollInterval" title="Polling interval" onSelect={this.onPollIntervalChanged}>
            {MetricsOptionsBar.PollIntervals.map(r => (
              <MenuItem key={r[0]} active={r[0] === this.state.pollInterval} eventKey={r[0]}>
                {r[1]}
              </MenuItem>
            ))}
          </DropdownButton>
        </div>
        <ToolbarRightContent>
          {this.props.loading && (
            <span>
              <Icon name="spinner" spin={true} size="lg" /> Loading
            </span>
          )}
        </ToolbarRightContent>
        {this.groupByLabelsHelper.hasResults() && (
          <Toolbar.Results>{this.groupByLabelsHelper.renderResults()}</Toolbar.Results>
        )}
      </Toolbar>
    );
  }
}

export default MetricsOptionsBar;
