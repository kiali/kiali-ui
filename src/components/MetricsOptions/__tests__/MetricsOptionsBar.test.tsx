import * as React from 'react';
import { mount, shallow } from 'enzyme';

import MetricsOptionsBar from '../MetricsOptionsBar';

const optionsChanged = jest.fn();
const lastOptionsChanged = () => {
  return optionsChanged.mock.calls[optionsChanged.mock.calls.length - 1][0];
};

describe('MetricsOptionsBar', () => {
  it('renders initial layout', () => {
    const wrapper = shallow(
      <MetricsOptionsBar onOptionsChanged={jest.fn()} onPollIntervalChanged={jest.fn()} onManualRefresh={jest.fn()} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('changes trigger parent callback', () => {
    const wrapper = mount(
      <MetricsOptionsBar
        onOptionsChanged={optionsChanged}
        onPollIntervalChanged={jest.fn()}
        onManualRefresh={jest.fn()}
      />
    );
    expect(optionsChanged).toHaveBeenCalledTimes(1);
    const opts = lastOptionsChanged();
    // Step = duration / ticks
    expect(opts).toHaveProperty('duration', MetricsOptionsBar.DefaultDuration);

    let elt = wrapper
      .find('#duration')
      .find('SafeAnchor')
      .first();
    elt.simulate('click');
    expect(optionsChanged).toHaveBeenCalledTimes(2);
    const expectedDuration = Object.keys(MetricsOptionsBar.Durations)[0];
    expect(lastOptionsChanged()).toHaveProperty('duration', expectedDuration);
  });
});
