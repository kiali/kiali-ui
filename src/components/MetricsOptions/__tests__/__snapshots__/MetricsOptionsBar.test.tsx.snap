// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetricsOptionsBar renders initial layout 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MetricsOptionsBar
    onManualRefresh={[MockFunction]}
    onOptionsChanged={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "byLabelsIn": Array [],
              "byLabelsOut": Array [],
              "duration": 60,
            },
          ],
        ],
      }
    }
    onPollIntervalChanged={
      [MockFunction] {
        "calls": Array [
          Array [
            15000,
          ],
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <getContext(Filter)>
          <FilterValueSelector
            className=""
            currentValue=""
            filterValues={
              Array [
                "local version",
                "remote service",
                "remote version",
                "response code",
              ]
            }
            id=""
            onFilterValueSelected={[Function]}
            placeholder="Group by"
          />
        </getContext(Filter)>,
        <ToolbarDropdown
          disabled={false}
          handleSelect={[Function]}
          id="metrics_filter_interval_duration"
          initialLabel="Last minute"
          initialValue={60}
          nameDropdown="Duration"
          options={
            Object {
              "10800": "Last 3 hours",
              "1800": "Last 30 minutes",
              "21600": "Last 6 hours",
              "2592000": "Last 30 days",
              "300": "Last 5 minutes",
              "3600": "Last hour",
              "43200": "Last 12 hours",
              "60": "Last minute",
              "600": "Last 10 minutes",
              "604800": "Last 7 days",
              "86400": "Last day",
            }
          }
        />,
        <ToolbarDropdown
          disabled={false}
          handleSelect={[Function]}
          id="metrics_filter_poll_interval"
          initialLabel="15 seconds"
          initialValue={15000}
          nameDropdown="Poll Interval"
          options={
            Object {
              "0": "Pause",
              "10000": "10 seconds",
              "15000": "15 seconds",
              "30000": "30 seconds",
              "300000": "5 minutes",
              "5000": "5 seconds",
              "60000": "1 minute",
            }
          }
        />,
        <ToolbarRightContent
          className=""
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={false}
            onClick={[MockFunction]}
          >
            <Icon
              name="refresh"
              type="fa"
            />
          </Button>
        </ToolbarRightContent>,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <FilterValueSelector
            className=""
            currentValue=""
            filterValues={
              Array [
                "local version",
                "remote service",
                "remote version",
                "response code",
              ]
            }
            id=""
            onFilterValueSelected={[Function]}
            placeholder="Group by"
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "",
            "currentValue": "",
            "filterValues": Array [
              "local version",
              "remote service",
              "remote version",
              "response code",
            ],
            "id": "",
            "onFilterValueSelected": [Function],
            "placeholder": "Group by",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": false,
          "handleSelect": [Function],
          "id": "metrics_filter_interval_duration",
          "initialLabel": "Last minute",
          "initialValue": 60,
          "nameDropdown": "Duration",
          "options": Object {
            "10800": "Last 3 hours",
            "1800": "Last 30 minutes",
            "21600": "Last 6 hours",
            "2592000": "Last 30 days",
            "300": "Last 5 minutes",
            "3600": "Last hour",
            "43200": "Last 12 hours",
            "60": "Last minute",
            "600": "Last 10 minutes",
            "604800": "Last 7 days",
            "86400": "Last day",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": false,
          "handleSelect": [Function],
          "id": "metrics_filter_poll_interval",
          "initialLabel": "15 seconds",
          "initialValue": 15000,
          "nameDropdown": "Poll Interval",
          "options": Object {
            "0": "Pause",
            "10000": "10 seconds",
            "15000": "15 seconds",
            "30000": "30 seconds",
            "300000": "5 minutes",
            "5000": "5 seconds",
            "60000": "1 minute",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={false}
            onClick={[MockFunction]}
          >
            <Icon
              name="refresh"
              type="fa"
            />
          </Button>,
          "className": "",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "active": false,
            "block": false,
            "bsClass": "btn",
            "bsStyle": "default",
            "children": <Icon
              name="refresh"
              type="fa"
            />,
            "disabled": false,
            "onClick": [MockFunction],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "name": "refresh",
              "type": "fa",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        "type": [Function],
      },
      false,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <getContext(Filter)>
            <FilterValueSelector
              className=""
              currentValue=""
              filterValues={
                Array [
                  "local version",
                  "remote service",
                  "remote version",
                  "response code",
                ]
              }
              id=""
              onFilterValueSelected={[Function]}
              placeholder="Group by"
            />
          </getContext(Filter)>,
          <ToolbarDropdown
            disabled={false}
            handleSelect={[Function]}
            id="metrics_filter_interval_duration"
            initialLabel="Last minute"
            initialValue={60}
            nameDropdown="Duration"
            options={
              Object {
                "10800": "Last 3 hours",
                "1800": "Last 30 minutes",
                "21600": "Last 6 hours",
                "2592000": "Last 30 days",
                "300": "Last 5 minutes",
                "3600": "Last hour",
                "43200": "Last 12 hours",
                "60": "Last minute",
                "600": "Last 10 minutes",
                "604800": "Last 7 days",
                "86400": "Last day",
              }
            }
          />,
          <ToolbarDropdown
            disabled={false}
            handleSelect={[Function]}
            id="metrics_filter_poll_interval"
            initialLabel="15 seconds"
            initialValue={15000}
            nameDropdown="Poll Interval"
            options={
              Object {
                "0": "Pause",
                "10000": "10 seconds",
                "15000": "15 seconds",
                "30000": "30 seconds",
                "300000": "5 minutes",
                "5000": "5 seconds",
                "60000": "1 minute",
              }
            }
          />,
          <ToolbarRightContent
            className=""
          >
            <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="default"
              disabled={false}
              onClick={[MockFunction]}
            >
              <Icon
                name="refresh"
                type="fa"
              />
            </Button>
          </ToolbarRightContent>,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <FilterValueSelector
              className=""
              currentValue=""
              filterValues={
                Array [
                  "local version",
                  "remote service",
                  "remote version",
                  "response code",
                ]
              }
              id=""
              onFilterValueSelected={[Function]}
              placeholder="Group by"
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "className": "",
              "currentValue": "",
              "filterValues": Array [
                "local version",
                "remote service",
                "remote version",
                "response code",
              ],
              "id": "",
              "onFilterValueSelected": [Function],
              "placeholder": "Group by",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": false,
            "handleSelect": [Function],
            "id": "metrics_filter_interval_duration",
            "initialLabel": "Last minute",
            "initialValue": 60,
            "nameDropdown": "Duration",
            "options": Object {
              "10800": "Last 3 hours",
              "1800": "Last 30 minutes",
              "21600": "Last 6 hours",
              "2592000": "Last 30 days",
              "300": "Last 5 minutes",
              "3600": "Last hour",
              "43200": "Last 12 hours",
              "60": "Last minute",
              "600": "Last 10 minutes",
              "604800": "Last 7 days",
              "86400": "Last day",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": false,
            "handleSelect": [Function],
            "id": "metrics_filter_poll_interval",
            "initialLabel": "15 seconds",
            "initialValue": 15000,
            "nameDropdown": "Poll Interval",
            "options": Object {
              "0": "Pause",
              "10000": "10 seconds",
              "15000": "15 seconds",
              "30000": "30 seconds",
              "300000": "5 minutes",
              "5000": "5 seconds",
              "60000": "1 minute",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="default"
              disabled={false}
              onClick={[MockFunction]}
            >
              <Icon
                name="refresh"
                type="fa"
              />
            </Button>,
            "className": "",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "active": false,
              "block": false,
              "bsClass": "btn",
              "bsStyle": "default",
              "children": <Icon
                name="refresh"
                type="fa"
              />,
              "disabled": false,
              "onClick": [MockFunction],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "name": "refresh",
                "type": "fa",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          "type": [Function],
        },
        false,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
